clinica/iotools/converters/adni_to_bids/adni_modalities/adni_fmap.py

```diff
@@ -245,165 +252,220 @@ def fmap_image(
+    if old_extension == "e1":
+        new_extension = "magnitude1"
+    elif old_extension == "e2":
+        new_extension = "magnitude2"
+    elif old_extension == "e1_ph":
+        new_extension = "phase1"
+    elif old_extension == "e2_ph":
+        new_extension = "phase2"
+    return new_extension
+
+
+def case2_2phase_2mag(fmap_path: Path):
+    "Performs the checks and renaming for BIDS spec case 2 for fieldmaps"
+    files = [f for f in os.listdir(fmap_path) if not f.startswith(".")]
+    check_json = [f for f in files if "ph.json" in f]
+    # Checking for Echotime keys in phase jsons
+    for js in check_json:
+        with open(fmap_path / js, "r") as file:
+            json_data = json.load(file)
+        if "EchoTime" not in json_data:
+            cprint(f'Invalid file {js}, missing "EchoTime" key.', lvl="warning")
+            unrecognized_fmap_case(fmap_path)
+            return
+    # Renaming
+    pattern = r"(.*_)fmap_(.*?)(\..*)"
+    for previous_filename in files:
+        rgx = re.search(pattern, previous_filename)
+        cut, extension, type = rgx.group(1), rgx.group(2), rgx.group(3)
+        new_name = cut + renaming_fmap_extensions_case2(extension) + type
+        os.rename(fmap_path / previous_filename, fmap_path / new_name)
+
+
+def renaming_fmap_extensions_case3(old_extension: str) -> str:
+    if old_extension == "":
+        new_extension = "fieldmap"
+    else:
+        new_extension = "magnitude"
+    return new_extension
+
+
+def direct_fieldmap(fmap_path: Path):
+    """Performs the checks and renaming for BIDS spec case 3 for fieldmaps"""
+
+    files = [f for f in os.listdir(fmap_path) if not f.startswith(".")]
+    # Assuming filename ends with "fmap" for the fieldmap
+    js = [f for f in files if f.endswith("fmap.json")][0]
+
+    # Checking for 'Units' key in fmap json
+    with open(fmap_path / js, "r") as file:
+        json_data = json.load(file)
+    if "Units" not in json_data:
+        cprint(
+            f'Invalid file {js} for Direct Fieldmapping, missing "Units" key.',
+            lvl="warning",
+        )
+        unrecognized_fmap_case(fmap_path)
+        return
+    # Renaming
+    pattern = r"(.*_)fmap(.*?)(\..*)"
+    for previous_filename in files:
+        rgx = re.search(pattern, previous_filename)
+        cut, extension, type = rgx.group(1), rgx.group(2), rgx.group(3)
+        new_name = cut + renaming_fmap_extensions_case3(extension) + type
+        os.rename(fmap_path / previous_filename, fmap_path / new_name)
+
+
+def unrecognized_fmap_case(fmap_path: Path):
+    """Deletes fmap directory"""
+    to_delete = os.listdir(fmap_path)
+    cprint(
+        f"The following files were found in {fmap_path} : {to_delete}."
+        f"They will be deleted as they are not usable.",
+        lvl="info",
+    )
+    for file in to_delete:
+        os.remove(fmap_path / file)
+
+
+def check_case_fmap(fmap_path: Path) -> str or None:
```

Comment:
This function infers which BIDS case we are dealing with. Depending on this inference, it returns either a string or None. This return value is then used as a key in the `bids_case` dictionary if it is a string in order to get a function to handle the specific case, and nothing happens if it is None.

This totally works but I would vote for a slightly more robust pattern. First, we can model all of these cases as an enumeration rather than using plain strings and Nones, something like that:

```python
class BIDSFMAPCase(int, Enum):  # used int but could be auto as well since we don't really need the value
    EMPTY_FOLDER = 0
    ONE_PHASE_TWO_MAGNITUDES = 1
    TWO_PHASES_TWO_MAGNITUDES = 2
    DIRECT_FIELDMAPS = 3
    NOT_SUPPORTED = 4
```

Then, you could have a factory for your handlers. Its only responsibility is to return the handler corresponding to the case passed as an input:

```python
def fmap_case_handler_factory(case: BIDSFMAPCase) -> Callable:
    if case == BIDSFMAPCase.EMPTY_FOLDER:
        return empty_folder_handler
    if case == BIDSFMAPCase.ONE_PHASE_TWO_MAGNITUDES:
        return one_phase_two_magnitudes_handler
    ...
```

All handlers must implement the same interface:

```python
def handler_name(fmap_path: Path):
    ...
```

This means that the empty handler would be something like that:

```python
def empty_folder_handler(fmap_path: Path):
    pass  # could also warn the user for example
```

The function `check_case_fmap` (`infer_case_fmap` maybe ?) would then have the following signature:

```python
def infer_case_fmap(fmap_path: Path) -> BIDSFMAPCase:
    ...
```

And return a variant of the `BIDSFMAPCase` enumeration. Finally, your entry point, i.e. the function `reorganize_fmap` would do something like that:

```python
for subject...:
    for session...:
        fmap_case_handler_factory(infer_case_fmap(fmap_path))(fmap_path)
```

WDYT ?

Reply:
Thanks @NicolasGensollen !! I was not sure how to do it cleanly, it seems much better like this.

Reply:
I updated the PR with your suggestions. The only change I made was using StrEnum with auto to be more explicit in case at some point it becomes interesting to look at the value